### PR TITLE
'hist': Montonic Constraints

### PR DESCRIPTION
**Pull request by redditur**
I have extended the monotonic constraints functionality so that they are now compatible with the 'hist' tree method. I have also added tests based on the repo by @madrury (https://github.com/madrury/xgboost-monotonic-testing).

### CODE REVIEW

**Comment by aldanor** [comment]
Looks like the build failed for unrelated reasons.

@redditur If you squash your commits (or just the most recent one) and force-push the branch it will retrigger the travis build.

**Comment by aldanor** [comment]
You can also [close the PR and then open it again](https://github.com/travis-ci/travis-ci/issues/576#issuecomment-65317915).

**Comment by aldanor** [comment]
R build failed, again for unrelated reasons.

Could this be possibly merged?

(@RAMitchell? Among other things, this fixes a bug introduced when monotonic constraints for gpu_hist were added and `CalcSplitGain()` signature was changed)

**Comment by redditur** [comment]
Yeah, it's related to a recent issue discussed on stackoverflow;(https://stackoverflow.com/questions/48545650/cannot-install-devtools-properly-without-digest-r)
As far as I can tell it should be a simple fix- editing the travis build config.

File: src/tree/param.h [hunk 1]
```diff
@@ -429,9 +430,9 @@ template <typename param_t>
     if (constraint == 0) {
       return gain;
     } else if (constraint > 0) {
-      return wleft < wright ? gain : 0.0;
+      return wleft <= wright ? gain : negative_infinity;
     } else {
-      return wleft > wright ? gain : 0.0;
+      return wleft >= wright ? gain : negative_infinity;
```
Comment (khotilov) [inline src/tree/param.h]: What is a reason for using -Inf here?

Comment (redditur) [reply src/tree/param.h]: To ensure that the split never happens.

Comment (khotilov) [reply src/tree/param.h]: I understand the general idea, but I'm curious why zero is not sufficient now, and whether there might potentially be any corner cases when the negative value might matter.

Comment (redditur) [reply src/tree/param.h]: As far as I understand- and I may be completely wrong- the split is calculated on the basis of the change in loss (loss_chg). You can see here;

(https://github.com/dmlc/xgboost/blob/master/src/tree/updater_fast_hist.cc#L869)

that the loss change (at least in the fast histogram updater) is calculated as the gain of the split minus the gain of the current node, hence the loss change can be negative.

Therefore, the negative infinity ensures that the loss change is as small as possible for any split which violates the constraints.

Comment (khotilov) [reply src/tree/param.h]: My memory could be wrong, but I thought that the negative loss change wasn't allowed by design.
However, if using -Inf doesn't break anything I'm ok with that, except please make the shortcut defined as const.

And I'm not sure whether that java_test fail is real or if travis is being a flaky primadonna again @superbobry ?

**Comment by redditur** [comment]
There's no obvious reason to me why this should break the code. 

Okay, that makes sense. I'll set the shortcut to a constant.

File: src/tree/param.h (hunk 1 above)
Comment (redditur) [reply src/tree/param.h]: There's no obvious reason to me why this should break the code.

Okay, that makes sense. I'll set the shortcut to a constant.

Comment (khotilov) [reply src/tree/param.h]: @redditur were you going to set the const?

**Comment by khotilov** [comment]
@hcho3 Does the approach in this PR look good to you?

**Comment by hcho3** [comment]
@khotilov I just got back from a conference (SysML 2018). I will write a review in a day or two.

File: src/tree/param.h (hunk 1 above)
Comment (hcho3) [reply src/tree/param.h]: @redditur Let me argue why the change is not strictly necessary, in terms of algorithmic correctness. The functions `CalcGain` and `CalcGainGivenWeight` always return nonnegative values by design. With algebra we can show that
```
Gain = G^2 / (H + lambda)
```
where G and H are the sums of first- and second-order gradients, respectively. Since all second-order gradients are nonnegative, H is also nonnegative and so is the gain. Since all gains are nonnegative, assigning gain 0 for invalid splits is okay.

At this point, we are arguing about aesthetics rather than correctness. The fact I just described is implicit and not well documented in code. So I'd still include `-Inf` for the sake of clarity.

Comment (hcho3) [reply src/tree/param.h]: @redditur And maybe I'm nitpicking here, but I'd use
```cpp
return wleft > wright ? gain : negative_infinity
```
instead of
```cpp
return wleft >= wright ? gain : negative_infinity
```
The existing code invalidates the split whenever `wleft == wright`, so we should honor this behavior.

**Comment by hcho3** [comment]
@aldanor When you said
> Among other things, this fixes a bug introduced when monotonic constraints for gpu_hist were added and CalcSplitGain() signature was changed)
Which bug are you referring to?

**Comment by hcho3** [comment]
@khotilov It looks good to me. We should merge it in once a few comments are addressed.

**Comment by hcho3** [comment]
@redditur Let me know whether you agree with our comments. If you do, I can make the necessary changes and merge it.

**Comment by aldanor** [comment]
@hcho3 

https://github.com/redditur/xgboost/blob/220e6d9dd0affb61d87154470acc593e7f424b6a/src/tree/updater_fast_hist.cc#L870

When `gpu_hist` was implemented, the `CalcSplitGain()` signature was changed (here: https://github.com/dmlc/xgboost/pull/2904/files), however not all of its uses were updated in the coded. Specifically, in `updater_fast_hist.cc`, the `fid` index was wrongfully passed in place of a monotonic constraint.

**Comment by aldanor** [comment]
@hcho3 

> The functions CalcGain and CalcGainGivenWeight always return nonnegative values by design

IIRC, the problem was that because of how `loss_chg` is calculated (split-gain minus gain), just setting the split-gain to 0 is not sufficient to prevent it from being used in case monotonic constraints are violated. As such, it's not an 'aesthetical' but rather a functional fix, @redditur may have some actual examples of that happening.

**Comment by hcho3** [comment]
@aldanor 
> When monotonic constraints support for gpu_hist was implemented, the CalcSplitGain() signature was changed (here: https://github.com/dmlc/xgboost/pull/2904/files), however not all of its uses were updated in the code. Specifically, in `updater_fast_hist.cc`, the `fid` index was wrongfully passed in place of a monotonic constraint.

I notice that`NoConstraint::CalcSplitGain()` and `ValueConstraint::CalcSplitGain()` now have different signatures. The code still ended up working so far because 1) the hist updater only supported `NoConstraint` and 2) `NoConstraint::CalcSplitGain()` does not use the second argument at all.

Now that we are supporting `ValueConstraint` in the hist updater, we should fix the signature for `NoConstraint::CalcSplitGain()`.

**Comment by hcho3** [comment]
@aldanor 
> just setting the split-gain to 0 is not sufficient to prevent it from being used in case monotonic constraints are violated

You would be right if there exists a case where all candidate splits have the split-gain of 0 and thus an invalid split (which also has split-gain of 0) may be chosen.

**Comment by khotilov** [comment]
@hcho3 The discussion about the loss change positivity reminded me about seeing some sporadic `assert self.non_decreasing(res['train']['auc'])` failures in tests/python/test_fast_hist.py. Setting `np.random.seed(4)` in there and then running with `nosetests  tests/python/test_fast_hist.py` seems to make it reproducible. I didn't do any more digging, but I thought there might be some link with this.

As a side-note: many python tests do not fix or use the numpy random generator state properly.

**Comment by hcho3** [comment]
@khotilov I had previously thought that the `min_split_loss` parameter would prevent negative loss change, but I was wrong. What's really happening is that the hist updater is allowed to make splits with negative loss change, and then later the pruner prunes the trees post-hoc using `min_split_loss` criteria. So for the purpose of monotonic constraints, we must not allow invalid splits at all, by setting negative infinity as the loss gain.

**Comment by aldanor** [comment]
> So for the purpose of monotonic constraints, we must not allow invalid splits at all, by setting negative infinity as the loss gain.

Yes, exactly.

**Comment by hcho3** [comment]
@aldanor I now stand corrected. We are indeed addressing a functional defect. One last thing: would you be in favor of replacing the line
```cpp
return wleft >= wright ? gain : negative_infinity
```
with
```cpp
return wleft > wright ? gain : negative_infinity
```
? I think we should preserve the behavior when `wleft == wright`.

**Comment by redditur** [comment]
>  @redditur may have some actual examples of that happening.

The tests I've written are sufficient examples to see this happening.

> we should fix the signature for NoConstraint::CalcSplitGain().

Okay, I'll do this.

> As a side-note: many python tests do not fix or use the numpy random generator state properly.

This is a side-issue, but I'm not comfortable with tests setting seeds; it's just asking for hackery to be built into the tests (i.e. people setting specific seeds so that their tests pass). A test should pass regardless of the seed, so why set it?

> What's really happening is that the hist updater is allowed to make splits with negative loss change, and then later the pruner prunes the trees post-hoc using min_split_loss criteria. 

This is my understanding, but, still, those splits should be pruned and shouldn't exist at the end of fitting, correct? Yet, this is not the case. Try the tests using 0 instead of -infinity and with a non-zero gamma and you'll see them failing sometimes. So, the pruning step doesn't quite work as one would expect.

> I think we should preserve the old behavior when wleft == wright.

Firstly, the old behaviour is wrong, so I'm not sure why it should be preserved. Secondly, I'm not sure what you mean by preserve the old behaviour? In the previous incarnation, when the weights were equal the CalcSplitGain method would have outputted 0. The code snippet

`return wleft > wright ? gain : negative_infinity`

on the other hand would output negative_infinity. So, at least in that sense, the behaviour wouldn't be preserved. Besides I think to anyone reading the code, the suggested code snippet looks strange;  why have a strict comparison?

**Comment by hcho3** [comment]
@redditur I meant to say that splits with `wleft == wright` should be invalidated.

**Comment by redditur** [comment]
> splits with wleft == wright should be invalidated

@hcho3 

What's the reasoning for this though? It would only make the learner stricter than it needs to be, which could hinder prediction performance unnecessarily.

**Comment by tqchen** [comment]
I think it is fine to remove constraint of for cases when ```wleft == wright```. In terms of loss comparison, it is better to return 0 as loss change can never be non-negative.

**Comment by redditur** [comment]
@tqchen 
> I think it is fine to remove constraint of for cases when wleft == wright

By which you mean keep the line 

`return wleft >= wright ? gain : negative_infinity`

as is?

> loss change can never be non-negative.

This is not true; see the above discussion.

**Comment by tqchen** [comment]
The point is that if the loss change ever becomes negative, the growing algorithm will choose not to split anyway, so setting things to 0 will be OK

**Comment by tqchen** [comment]
I do mean we can keep the equality

**Comment by hcho3** [comment]
>if the loss change ever becomes negative, the growing algorithm will choose not to split anyway, so setting things to 0 will be OK

I thought so too, but the parameter `min_split_loss` does not get applied at the time of tree construction.

**Comment by hcho3** [comment]
@aldanor @redditur I am almost convinced about merging this PR in. Just to make sure, would you be able to provide us with an example where the old behavior results in an invalid split? Thanks!

**Comment by redditur** [comment]
> Just to make sure, would you be able to provide us with an example where the old behavior results in an invalid split?

@hcho3 

Sure. Take the pull request as it is at the minute and change negative_infinity to 0. Then in the tests set numpy's seed,

`seed = 1234`
`np.random.seed(seed)`

and set the same seed in the parameters for the constrained_hist_method; that is,

`params_for_constrained_hist_method = {`
`    'tree_method': 'hist', 'silent': 1,`
`    'monotone_constraints': '(1, -1)', 'seed': seed`
`}`

The test for the constrained hist method will then fail.

**Comment by hcho3** [comment]
@redditur I managed to reproduce the bug for 'hist'. Thanks! Do you also have an example where the old behavior fails for 'exact' method?

**Comment by hcho3** [comment]
@redditur I looked into the code and found out that for some reason `root_gain` is being set to negative value. This phenomenon doesn't occur for the `exact` method. I'm trying to find out whether this behavior is legitimate or not.

**Comment by redditur** [comment]
> By any chance, do you also have an example where the old behavior fails for 'exact' method?

@hcho3 

No, I don't. I've never seen the 'exact' method fail a monotonicity constraint and so think this is a 'hist' method specific issue.

**Comment by hcho3** [comment]
Here is my findings, as for why the old behavior is broken:
* When monotonic constraints are applied, gain values may possibly be negative.
* Thus, it is possible that all valid splits get negative gains (@redditur gave us an example where this actually occurs); since invalid splits were assigned a gain of 0, an invalid split got picked as the "best".
So I am now in favor of merging @redditur 's PR.

Before we merge in this code, we must decide on one issue: how should we handle the scenario where all valid splits result in negative loss change?
1) Do not split at all
2) Choose the best valid split, even with a negative loss change.
Right now, this scenario is handled incorrectly: the split is set to `(feature 0, threshold 0.0)` because the accumulator variable `snode[nid].best` gets initialized to `(feature 0, threshold 0.0, loss_chg 0.0)` and never gets updated afterwards.

**Comment by khotilov** [comment]
@hcho3 1. Do not split at all sounds reasonable to me. The monotonicity requirement seems to act as a kind of regularization that might result in situations where making any split would be detrimental.

**Comment by hcho3** [comment]
@khotilov I agree. I'll make a change to reflect this decision.

**Comment by redditur** [comment]
> how should we handle the scenario where all valid splits result in negative loss change?

@hcho3 

I feel that we are in danger of this pull request diverging from its intended purpose. I think, therefore, that this should be opened as a separate issue. 

However, for what it's worth, I couldn't possibly decide between;

>1. Do not split at all
>2. Choose the best valid split, even with a negative loss change.

without understanding in more detail some of the decisions already taken in codebase.

Firstly, I don't understand what the loss change actually refers to and therefore don't understand why splits are compared based on this quantity. The name suggests that it refers to the change in some loss function brought about by a split, but the code doesn't suggest this is the case and, moreover, if this were the case, why would this quantity ever be negative?

Why aren't splits chosen based on the output of CalcSplitGain, which, if named appropriately, is the gain brought about by a proposed split?

**Comment by hcho3** [comment]
> I feel that we are in danger of this pull request diverging from its intended purpose.

The intended purposr of this PR is to support monotonic constraints in 'hist' method. Your addition revealed a corner case that was not properly handled before. Fixing this is crucial for correct behavior of monotonic constraints, so I think it's reasonable to deal with it here.

>The name suggests that it refers to the change in some loss function brought about by a split ... if
this were the case, why would this quantity ever be negative?

Indeed the name does suggest the change in the loss function brought about by the split. It is actually possible that this quantity be negative ... because the leaf weight is modified by the monotonic constraints. (Without monotonic constraints, you do get always positive loss change, but with then you don't.)

**Comment by hcho3** [comment]
@redditur Now that I think about it, I'm actually fine with opening another PR. Why don't you merge your work on top of latest master? And then I'll propose my bug fix. Thanks!

**Comment by hcho3** [comment]
@redditur I went ahead and deleted my addition from this PR. I'll merge it when tests pass.

**Comment by aldanor** [comment]
Looks like the tests pass ok?

**Comment by hcho3** [comment]
Done. Thank you all!
